lib/protocol/bufferpool_test.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2019 The Protocol Authors.
+
+package protocol
+
+import "testing"
+
+func TestBucketNumbers(t *testing.T) {
+	cases := []struct {
+		size int
+		put  int
+		get  int
+	}{
+		// Small blocks are put nowwhere, but can be fetched from bucket 0
+		{size: 1024, put: -1, get: 0},
+
+		// The blocksize for bucket zero  goes there
+		{size: MinBlockSize, put: 0, get: 0},
+
+		// Up to the next blocksize - 1 we still put in the same bucket, but
+		// we look for these blocks in the next bucket where we know they
+		// can be found.
+		{size: 2*MinBlockSize - 1, put: 0, get: 1},
+
+		// Next blocksize at the border...
+		{size: 2 * MinBlockSize, put: 1, get: 1},
+
+		// ... and past it.
+		{size: 2*MinBlockSize + 1, put: 1, get: 2},
```

Comment:
I don't get this. We should always allocate bucket sized slices, and we should always get and return them to the same pool.

Reply:
Granted the cap should be an even block size when we put it, but the size we are asking for will be random and the two needs to be handled separately anyway. Plus in this way we survive if someone puts a random buffer. 🤷‍♂

Reply:
We readjust the size on reading anyway, so we should put stuff based on cap, if it does match a bucket, panic.